sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -91,1012 +36,268 @@
-    }
-  }
-
-  public static Class<?> getDoFnClass(OldDoFn<?, ?> fn) {
-    if (fn instanceof SimpleDoFnAdapter) {
-      return ((SimpleDoFnAdapter<?, ?>) fn).fn.getClass();
-    } else {
-      return fn.getClass();
-    }
-  }
-
-  private static class SimpleDoFnAdapter<InputT, OutputT> extends OldDoFn<InputT, OutputT> {
-
-    private final DoFn<InputT, OutputT> fn;
-    private transient DoFnInvoker<InputT, OutputT> invoker;
-
-    private SimpleDoFnAdapter(DoFnReflector reflector, DoFn<InputT, OutputT> fn) {
-      super(fn.aggregators);
-      this.fn = fn;
-      this.invoker = reflector.bindInvoker(fn);
-    }
-
-    @Override
-    public void startBundle(OldDoFn<InputT, OutputT>.Context c) throws Exception {
-      ContextAdapter<InputT, OutputT> adapter = new ContextAdapter<>(fn, c);
-      invoker.invokeStartBundle(adapter, adapter);
-    }
-
-    @Override
-    public void finishBundle(OldDoFn<InputT, OutputT>.Context c) throws Exception {
-      ContextAdapter<InputT, OutputT> adapter = new ContextAdapter<>(fn, c);
-      invoker.invokeFinishBundle(adapter, adapter);
-    }
-
-    @Override
-    public void processElement(OldDoFn<InputT, OutputT>.ProcessContext c) throws Exception {
-      ProcessContextAdapter<InputT, OutputT> adapter = new ProcessContextAdapter<>(fn, c);
-      invoker.invokeProcessElement(adapter, adapter);
+  static DoFnSignature.BundleMethod analyzeBundleMethod(
+      Method m, TypeToken<?> inputT, TypeToken<?> outputT) {
+    if (!void.class.equals(m.getReturnType())) {
+      throw new IllegalArgumentException(
+          String.format("%s must have a void return type", format(m)));
     }
-
-    @Override
-    protected TypeDescriptor<InputT> getInputTypeDescriptor() {
-      return fn.getInputTypeDescriptor();
+    if (m.isVarArgs()) {
+      throw new IllegalArgumentException(String.format("%s must not have var args", format(m)));
     }
 
-    @Override
-    protected TypeDescriptor<OutputT> getOutputTypeDescriptor() {
-      return fn.getOutputTypeDescriptor();
-    }
+    TypeToken<?> expectedContextToken = doFnContextTypeOf(inputT, outputT);
 
-    @Override
-    public Duration getAllowedTimestampSkew() {
-      return fn.getAllowedTimestampSkew();
-    }
-
-    @Override
-    public void populateDisplayData(DisplayData.Builder builder) {
-      builder.include(fn);
+    Type[] params = m.getGenericParameterTypes();
+    if (params.length != 1) {
+      throw new IllegalArgumentException(
+          String.format(
+              "%s must have a single argument of type %s",
+              format(m), formatType(expectedContextToken)));
     }
-
-    private void readObject(java.io.ObjectInputStream in)
-        throws IOException, ClassNotFoundException {
-      in.defaultReadObject();
-      invoker = DoFnReflector.of(fn.getClass()).bindInvoker(fn);
+    TypeToken<?> contextToken = TypeToken.of(params[0]);
+    if (!contextToken.getRawType().equals(expectedContextToken.getRawType())) {
```

Comment:
Do we need to use raw types here? Could this be done as a check on the contextToken vs. the expectedContextToken?